libchisel/src/verifyinstructions.rs

```diff
@@ -0,0 +1,297 @@
+use super::InstructionValidator;
+
+use parity_wasm::elements::Instruction;
+use parity_wasm::elements::Instruction::*;
+use parity_wasm::elements::*;
+
+use crate::instructionerrors::*;
+
+pub const GET_INST: [Instruction; 2] = [GetGlobal(0), GetLocal(0)];
+
+pub const I32_BINOP: [Instruction; 15] = [
+    I32Add, I32Sub, I32Mul, I32DivS, I32DivU, I32RemS, I32RemU, I32And, I32Or, I32Xor, I32Shl,
+    I32ShrS, I32ShrU, I32Rotl, I32Rotr,
+];
+
+pub const I64_BINOP: [Instruction; 15] = [
+    I64Add, I64Sub, I64Mul, I64DivS, I64DivU, I64RemS, I64RemU, I64And, I64Or, I64Xor, I64Shl,
+    I64ShrS, I64ShrU, I64Rotl, I64Rotr,
+];
+
+pub const F32_BINOP: [Instruction; 7] =
+    [F32Add, F32Sub, F32Mul, F32Div, F32Min, F32Max, F32Copysign];
+
+pub const F64_BINOP: [Instruction; 7] =
+    [F64Add, F64Sub, F64Mul, F64Div, F64Min, F64Max, F64Copysign];
+
+use self::Filter::*;
+use std::mem::discriminant;
+
+type Pop = Vec<ValueType>;
+type Push = Vec<ValueType>;
+
+struct Signature {
+    pop: Pop,
+    push: Push,
+}
+
+pub enum Filter {
+    NumericInstructions,
+    NoFilter,
+}
+
+/// Basic struct for validating modules
+pub struct VerifyInstructions {
+    filter: Filter,
+    stack: Vec<ValueType>,
+}
+
+impl VerifyInstructions {
+    pub fn new(filter: Filter) -> Self {
+        VerifyInstructions {
+            filter,
+            stack: vec![],
+        }
+    }
+
+    fn check_instructions(
+        &mut self,
+        module: &Module,
+        body: &FuncBody,
+        index: usize,
+    ) -> Result<bool, InstructionError> {
+        for instruction in body.code().elements() {
+            if contains(instruction, &GET_INST)
+                && !self.push_global_or_local(module, instruction, body, index)?
+            {
+                return Ok(false);
+            }
+            match self.filter {
+                NumericInstructions => {
+                    let signature = get_instruction_signature(instruction);
+                    // if the instruction does not have a signature we are interested in, we continue
+                    if signature.is_some()
+                        && !self.validate_instruction(&signature.unwrap(), instruction)?
+                    {
+                        return Ok(false);
+                    }
+                }
+                NoFilter => (), // TODO: do this
+            };
+        }
+        Ok(true)
+    }
+
+    fn validate_instruction(
+        &mut self,
+        signature: &Signature,
+        instruction: &Instruction,
+    ) -> Result<bool, InstructionError> {
+        for signature_value in &signature.pop {
+            let value = self.stack.pop();
+            match value {
+                Some(stack_value) => {
```

Comment:
It is a matter of preference but I find it cleaner, in the case of destructuring two-valued enums like `Option`, to use the `if let Some(stack_value)` pattern here.